File: runtimecs/java/util/Comparator.cs
using java.lang;

namespace java.util 
{
    public interface Comparator
    {
        int compare(object o1, object o2);
        Comparator reversed();
        Comparator thenComparing(Comparator other);
    }

    public static class Comparator_0009
    {
        public static Comparator reversed(Comparator @this)
        {   
            if (@this==null) { throw new NullPointerException(); }
            return new ComparatorReversed(@this);
        }
        public static Comparator thenComparing(Comparator @this, Comparator other)
        {   
            if (@this==null || other==null) { throw new NullPointerException(); }
            return new ComparatorThenComparing(@this, other);
        }
    }

    internal class ComparatorReversed: Comparator
    {
        protected readonly Comparator other;
        
        public ComparatorReversed(Comparator other) 
        {   
            this.other = other;
        }
        public virtual int compare(object o1, object o2)
        {   
            return other.compare(o2,o1);
        }            
        public virtual Comparator thenComparing(Comparator other)
        {   
            return Comparator_0009.thenComparing(this,other);
        }    
        public virtual Comparator reversed()
        {   
            return other;
        }    
    }
            
    internal class ComparatorThenComparing: ComparatorReversed
    {
        private readonly Comparator second;
        public ComparatorThenComparing(Comparator other, Comparator second) : base(other)
        { 
            this.second = second;
        }
        public override int compare(object o1, object o2)
        {   
            int v = other.compare(o1,o2);
            if (v!=0) { return v; }
            return second.compare(o1,o2);
        }    
    }
}

File: runtimejs/java/util/Comparator.js
//load// java/lang/Object
var java_util_Comparator = { 
    _superinterfaces: null, 
    _defaults: {
        reversed_0: function()
        {   return new java_util_ComparatorReversed.$(this);
        },
        thenComparing_1: function(other)
        {   return new java_util_ComparatorThenComparing.$(this,other);
        },
    }
}; 

// -- methods:
// int compare(T o1, T o2)

var java_util_ComparatorReversed =
{   $: function(a)   // internal use only - integrate allocator with constructor 
    {   this.a = a;
    },
};
_class(java_util_ComparatorReversed, java_lang_Object, [java_util_Comparator], "java.util.ComparatorReversed", 
{   compare_2: function(x,y) 
    {   return -this.a.compare_2(x,y);
    },        
});

var java_util_ComparatorThenComparing =
{   $: function(a,b)   // internal use only - integrate allocator with constructor 
    {   this.a = a;
        this.b = b;
    },
};
_class(java_util_ComparatorThenComparing, java_lang_Object, [java_util_Comparator], "java.util.ComparatorThenComparing", 
{   compare_2: function(x,y) 
    {   var v = this.a.compare_2(x,y);
        if (v!=0) return v;
        return this.b.compare_2(x,y);
    },        
});
